File: lib/src/stdlib/stdlib.c
// NebulaOS - stdlib implementation
// ==================================
//
// Minimal standard library

#include "../../lib/include/stdlib.h"
#include "../../lib/include/string.h"
#include "../../kernel/common/include/nebula.h"

static unsigned int seed = 1;

int atoi(const char* str) {
    return (int)strtol(str, NULL, 10);
}

long atol(const char* str) {
    return strtol(str, NULL, 10);
}

long strtol(const char* str, char** endptr, int base) {
    const char* p = str;
    long result = 0;
    int sign = 1;
    
    while (*p == ' ' || *p == '\t') p++;
    
    if (*p == '-') { sign = -1; p++; }
    else if (*p == '+') p++;
    
    if (base == 0) {
        if (p[0] == '0') {
            if (p[1] == 'x' || p[1] == 'X') base = 16;
            else base = 8;
        } else base = 10;
    }
    
    while (*p) {
        int digit;
        if (*p >= '0' && *p <= '9') digit = *p - '0';
        else if (*p >= 'a' && *p <= 'z') digit = *p - 'a' + 10;
        else if (*p >= 'A' && *p <= 'Z') digit = *p - 'A' + 10;
        else break;
        
        if (digit >= base) break;
        
        result = result * base + digit;
        p++;
    }
    
    if (endptr) *endptr = (char*)p;
    return sign * result;
}

unsigned long strtoul(const char* str, char** endptr, int base) {
    const char* p = str;
    unsigned long result = 0;
    
    while (*p == ' ' || *p == '\t') p++;
    if (*p == '+') p++;
    
    if (base == 0) {
        if (p[0] == '0') {
            if (p[1] == 'x' || p[1] == 'X') base = 16;
            else base = 8;
        } else base = 10;
    }
    
    while (*p) {
        int digit;
        if (*p >= '0' && *p <= '9') digit = *p - '0';
        else if (*p >= 'a' && *p <= 'z') digit = *p - 'a' + 10;
        else if (*p >= 'A' && *p <= 'Z') digit = *p - 'A' + 10;
        else break;
        
        if (digit >= base) break;
        result = result * base + digit;
        p++;
    }
    
    if (endptr) *endptr = (char*)p;
    return result;
}

int abs(int x) {
    return x < 0 ? -x : x;
}

long labs(long x) {
    return x < 0 ? -x : x;
}

void srand(unsigned int s) {
    seed = s;
}

int rand(void) {
    seed = seed * 1103515245 + 12345;
    return (int)(seed / 65536) % 32768;
}

void qsort(void* base, size_t nmemb, size_t size, int (*compar)(const void*, const void*)) {
    if (nmemb < 2) return;
    
    char* arr = (char*)base;
    for (size_t i = 0; i < nmemb - 1; i++) {
        for (size_t j = 0; j < nmemb - i - 1; j++) {
            if (compar(arr + j * size, arr + (j + 1) * size) > 0) {
                char tmp[256];
                memcpy(tmp, arr + j * size, size);
                memcpy(arr + j * size, arr + (j + 1) * size, size);
                memcpy(arr + (j + 1) * size, tmp, size);
            }
        }
    }
}

void* bsearch(const void* key, const void* base, size_t nmemb, size_t size, int (*compar)(const void*, const void*)) {
    size_t low = 0, high = nmemb;
    const char* arr = (const char*)base;
    
    while (low < high) {
        size_t mid = low + (high - low) / 2;
        int cmp = compar(key, arr + mid * size);
        if (cmp == 0) return (void*)(arr + mid * size);
        else if (cmp < 0) high = mid;
        else low = mid + 1;
    }
    
    return NULL;
}

void exit(int status) {
    (void)status;
    kernel_panic("exit() called");
}

File: lib/src/stdlib.rs
pub unsafe fn atoi(s: *const u8) -> i32 {
    let mut i = 0;
    let mut sign = 1;
    let mut result = 0;
    if *s == b'-' {
        sign = -1;
        i += 1;
    }
    while *s.add(i) >= b'0' && *s.add(i) <= b'9' {
        result = result * 10 + (*s.add(i) - b'0') as i32;
        i += 1;
    }
    result * sign
}

pub unsafe fn itoa(value: i32, buffer: *mut u8) {
    let mut i = 0;
    let mut val = value;
    if val == 0 {
        *buffer = b'0';
        *buffer.add(1) = 0;
        return;
    }
    if val < 0 {
        *buffer.add(i) = b'-';
        i += 1;
        val = -val;
    }
    let mut start = i;
    while val > 0 {
        *buffer.add(i) = (val % 10) as u8 + b'0';
        i += 1;
        val /= 10;
    }
    let mut j = start;
    let mut k = i - 1;
    while j < k {
        let tmp = *buffer.add(j);
        *buffer.add(j) = *buffer.add(k);
        *buffer.add(k) = tmp;
        j += 1;
        k -= 1;
    }
    *buffer.add(i) = 0;
}

static mut RANDOM_SEED: u32 = 1;

pub unsafe fn srand(seed: u32) {
    RANDOM_SEED = seed;
}

pub unsafe fn rand() -> i32 {
    RANDOM_SEED = RANDOM_SEED.wrapping_mul(1103515245).wrapping_add(12345);
    ((RANDOM_SEED / 65536) % 2048) as i32
}
